src/mscorlib/src/System/Buffer.cs

```diff
@@ -480,51 +480,44 @@ internal unsafe static void Memmove(byte* dest, byte* src, nuint len)
                 *(int*)(dest + 16) = *(int*)(src + 16);
                 *(short*)(dest + 20) = *(short*)(src + 20);
                 return;
+            case 23:
+#if BIT64
+                *(long*)dest = *(long*)src;
+                *(long*)(dest + 8) = *(long*)(src + 8);
+#else
+                *(int*)dest = *(int*)src;
+                *(int*)(dest + 4) = *(int*)(src + 4);
+                *(int*)(dest + 8) = *(int*)(src + 8);
+                *(int*)(dest + 12) = *(int*)(src + 12);
+#endif
+                *(int*)(dest + 16) = *(int*)(src + 16);
+                *(short*)(dest + 20) = *(short*)(src + 20);
+                *(dest + 22) = *(src + 22);
+                return;
             }
 
             // P/Invoke into the native version for large lengths
             if (len >= 512) goto PInvoke;
-
-            nuint i = 0; // byte offset at which we're copying
-
-            if (((int)dest & 3) != 0)
-            {
-                if (((int)dest & 1) != 0)
-                {
-                    *(dest + i) = *(src + i);
-                    i += 1;
-                    if (((int)dest & 2) != 0)
-                        goto IntAligned;
-                }
-                *(short*)(dest + i) = *(short*)(src + i);
-                i += 2;
-            }
-
-            IntAligned:
-
-#if BIT64
-            // On 64-bit IntPtr.Size == 8, so we want to advance to the next 8-aligned address. If
-            // (int)dest % 8 is 0, 5, 6, or 7, we will already have advanced by 0, 3, 2, or 1
-            // bytes to the next aligned address (respectively), so do nothing. On the other hand,
-            // if it is 1, 2, 3, or 4 we will want to copy-and-advance another 4 bytes until
-            // we're aligned.
-            // The thing 1, 2, 3, and 4 have in common that the others don't is that if you
-            // subtract one from them, their 3rd lsb will not be set. Hence, the below check.
-
-            if ((((int)dest - 1) & 4) == 0)
-            {
-                *(int*)(dest + i) = *(int*)(src + i);
-                i += 4;
-            }
-#endif // BIT64
-
+            
+            int alignment = IntPtr.Size - 1;
+
+            // First write: copy a single word from src to dest
+            // We don't care if this is aligned; if it isn't, we'll just start
+            // at the next aligned address and maybe re-copy a few bytes. The
+            // benefit of this is it allows us to avoid multiple branches and
+            // reduces code size.
+            *(nuint*)dest = *(nuint*)src;
+
+            // byte offset at which we're copying
+            nuint i = 8u - ((nuint)dest & (nuint)alignment); // maps 0 => 8, 1 => 7, ... 8 => 8 for 64-bit
```

Comment:
@mikedn With `int` the compiler has to [treat the sign bit specially](http://tryroslyn.azurewebsites.net/#b:master/K4Zwlgdg5gBAygTxAFwKYFsDcAoADsAIwBswBjGUogQxBBgGEYBvbGNmfYsmANwHswAExgBZABQBKZq3az6fCCD5FUAOgDqAJzBoAMpFRiAtAEYYAMhgB2CTllyFSlRu16DxswFIYADlsy2AF9sQKA==), so we end up with this (very inefficient) code if we use `%` rather than `&`:

```asm
G_M6842_IG27:
       4981F800020000       cmp      r8, 512
       736C                 jae      SHORT G_M6842_IG30
       488B02               mov      rax, qword ptr [rdx]
       488901               mov      qword ptr [rcx], rax
       8BC1                 mov      eax, ecx
       448BC8               mov      r9d, eax
       41C1F91F             sar      r9d, 31
       4183E107             and      r9d, 7
       4403C8               add      r9d, eax
       4183E1F8             and      r9d, -8
       412BC1               sub      eax, r9d
       F7D8                 neg      eax
       83C008               add      eax, 8
       4863C0               movsxd   rax, eax
       4D8D48F0             lea      r9, [r8-16]
       4D8BD0               mov      r10, r8
       4C2BD0               sub      r10, rax
```